File: server/src/main/java/dev/sophiawhite/level/Property.java
package dev.sophiawhite.level;

import com.fasterxml.jackson.databind.JsonNode;

import java.util.ArrayList;
import java.util.List;

public class Property {

    private String name;
    private String type;
    private String value;

    public Property(String name, String type, String value) {
        this.name = name;
        this.type = type;
        this.value = value;
    }

    public String getName(){
        return this.name;
    }

    public String getType(){
        return this.type;
    }

    public String getValue(){
        return this.value;
    }

    public static Property singlePropertyFromJson(JsonNode property) {
        String name = property.path("name").asText();
        String type = property.path("type").asText();
        String value = property.path("value").asText();
        return new Property(name, type, value);
    }

    /**
     *
     * @param properties Array OR single JSON object expected.
     * @return List of all properties in the array, or the single property provided if input was not array.
     */
    public static List<Property> fromJson(JsonNode properties) {

        List<Property> output = new ArrayList<Property>();

        if (properties.isArray()) {
            for (JsonNode property : properties) {
                output.add(singlePropertyFromJson(property));
            }
        } else {
            output.add(singlePropertyFromJson(properties));
        }

        return output;
    }


}

File: client/script/level/Property.js
export class Property {

    /** @type {string} */ #name;
    /** @type {string} */ #type;
    /** @type {string} */ #value;

    /**
     * @param {string} name
     * @param {string} type
     * @param {string} value
     */
    constructor(name, type, value) {
        this.#name = name;
        this.#type = type;
        this.#value = value;
    }

    get name() {
        return this.#name;
    }
    get type() {
        return this.#type;
    }
    get value() {
        return this.#value;
    }

    toString(){
        return JSON.stringify({
            name: this.#name,
            type: this.#type,
            value: this.#value
        }, null, 2);
    }
  
}

export default Property;